NDTensors/src/blocksparse/blocksparsetensor.jl

```diff
@@ -335,11 +335,15 @@ view(T::BlockSparseTensor, b::Block) = blockview(T, b)
 # convert to Dense
 function dense(T::TensorT) where {TensorT<:BlockSparseTensor}
   R = zeros(dense(TensorT), inds(T))
+  ## Here this failed with scalar indexing (R[blockindices] = blockview)
+  ## We can fix this by using copyto the arrays
+  r = Array(R)
```

Comment:
Does this make a GPU array if `T` was on GPU? If so that seems like a bug since `Array` should only every output a CPU array and we should use something else. It's a bit of a hack but maybe just use `reshape(data(R), size(R))` for now.

Reply:
Or maybe use `array(R)`?

Reply:
Array does make a GPU array if T has a GPU backend. Yes I do feel like this code is a hack. Really we need a way to create an unallocated array of the correct size and shape, then we need to loop through every block. If the block is zero we fill the view with 0's, if the block is non-zero we use `copyto!`. After thats finished we can package it into `Tensor` and `Dense`

Reply:
I think it's not worth doing anything too sophisticated right now since I'm rewriting a lot of this anyway using the `BlockSparseArray` type, but I also just don't want to use incorrect code and I consider `r = Array(R)` to be incorrect (i.e. it is a bug that it outputs a GPU array).

So does it work to use `r = array(R)` here?

Reply:
Though for the record I disagree with your suggestion, I think it is best to write a conversion from a block sparse array to a dense array by first allocating a zero array of the full size, then looping through just the nonzero blocks and slicing them over. If the block sparse array is very sparse that will be a lot less work than looping through every block (imagine the limit where the block sizes are really small and there are many blocks). That's the strategy here and will also be how the new code is designed. It is pretty fast to allocate all of the zeros up front.

Reply:
(There may be some crossover where your suggestion is faster for a small number of large blocks since you're not allocating nonzero blocks twice, perhaps worth investigating but this operation is likely not a bottleneck anywhere anyway.)

Reply:
I agree, I don't think that this operation is the bottleneck for any of our algorithms at the moment but it definitely would be something interesting to look into!

Reply:
I completely understand your reasoning for seeing `r = Array(R)` as incorrect and agree. Also using `r = array(R)` does work so I will change the code to use `array`